core/node/rpc/sync/client/syncer_set.go

```diff
@@ -0,0 +1,268 @@
+		if nodeAddress == localNodeAddress { // stream managed by this node
+			syncer, err := newLocalSyncer(ctx, localNodeAddress, streamCache, cookieSet.AsSlice(), messages)
+			if err != nil {
+				return nil, nil, err
+			}
+			syncers[nodeAddress] = syncer
+		} else {
+			client, err := nodeRegistry.GetStreamServiceClientForAddress(nodeAddress)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			syncer, err := newRemoteSyncer(ctx, syncID, nodeAddress, client, cookieSet.AsSlice(), messages)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			syncers[nodeAddress] = syncer
+		}
+
+		// associate syncer with streamId to remove stream from sync operation
+		syncer := syncers[nodeAddress]
+		for streamID := range cookieSet {
+			streamID2Syncer[streamID] = syncer
+		}
+	}
+
+	return &SyncerSet{
+		ctx:              ctx,
+		syncID:           syncID,
+		streamCache:      streamCache,
+		nodeRegistry:     nodeRegistry,
+		localNodeAddress: localNodeAddress,
+		syncers:          syncers,
+		streamID2Syncer:  streamID2Syncer,
+		messages:         messages,
+	}, messages, nil
+}
+
+func (ss *SyncerSet) Run() {
+	ss.muSyncers.Lock()
+	ss.syncerTasks.Add(len(ss.syncers))
+	for syncerAddr, syncer := range ss.syncers {
+		go func() {
+			syncer.Run()
+			ss.muSyncers.Lock()
+			delete(ss.syncers, syncerAddr)
+			ss.muSyncers.Unlock()
+			ss.syncerTasks.Done()
+		}()
+	}
+	ss.muSyncers.Unlock()
+
+	<-ss.ctx.Done()       // sync cancelled by client or client conn dropped
+	ss.syncerTasks.Wait() // background syncers finished
+	close(ss.messages)    // close will cause the sync operation to send the SYNC_CLOSE message to the client
+}
+
+func (ss *SyncerSet) AddStream(ctx context.Context, nodeAddress common.Address, streamID StreamId, cookie *SyncCookie) error {
+	ss.muSyncers.Lock()
+	defer ss.muSyncers.Unlock()
+
+	if _, found := ss.streamID2Syncer[streamID]; found {
+		return nil // stream is already part of sync operation
+	}
+
+	// check if there is already a syncer that can sync the given stream -> add stream to the syncer
+	if syncer, found := ss.syncers[nodeAddress]; found {
+		if err := syncer.AddStream(ctx, cookie); err != nil {
+			return err
+		}
+		ss.streamID2Syncer[streamID] = syncer
+		return nil
+	}
+
+	// first stream to sync with remote -> create a new syncer instance
+	var (
+		syncer StreamsSyncer
+		err    error
+	)
+	if nodeAddress == ss.localNodeAddress {
+		if syncer, err = newLocalSyncer(ss.ctx, ss.localNodeAddress, ss.streamCache, []*SyncCookie{cookie}, ss.messages); err != nil {
+			return err
+		}
+	} else {
+		client, err := ss.nodeRegistry.GetStreamServiceClientForAddress(nodeAddress)
+		if err != nil {
+			return err
+		}
+		if syncer, err = newRemoteSyncer(ss.ctx, ss.syncID, nodeAddress, client, []*SyncCookie{cookie}, ss.messages); err != nil {
+			return err
+		}
+	}
+
+	ss.syncers[nodeAddress] = syncer
+	ss.streamID2Syncer[streamID] = syncer
+
+	ss.syncerTasks.Add(1)
+	go func() {
```

Comment:
looks like this can be a functions on ss, then here and above in can be go-called instead of same code.